test/e2e/gameserverallocation_test.go

```diff
@@ -129,6 +129,78 @@ func TestCreateFleetAndGameServerStateFilterAllocation(t *testing.T) {
+	if !runtime.FeatureEnabled(runtime.FeatureStateAllocationFilter) {
+		t.SkipNow()
+	}
+	t.Parallel()
+	log := e2e.TestLogger(t)
+	ctx := context.Background()
+
+	fleets := framework.AgonesClient.AgonesV1().Fleets(framework.Namespace)
+	fleet := defaultFleet(framework.Namespace)
+	lockLabel := "agones.dev/sdk-available"
+	// to start they are all available
+	fleet.Spec.Template.ObjectMeta.Labels = map[string]string{lockLabel: "true"}
+
+	flt, err := fleets.Create(ctx, fleet, metav1.CreateOptions{})
+	require.NoError(t, err)
+	defer fleets.Delete(ctx, flt.ObjectMeta.Name, metav1.DeleteOptions{}) // nolint:errcheck
+
+	framework.AssertFleetCondition(t, flt, e2e.FleetReadyCount(flt.Spec.Replicas))
+
+	fleetSelector := metav1.LabelSelector{MatchLabels: map[string]string{agonesv1.FleetNameLabel: flt.ObjectMeta.Name}}
+	allocatedSelector := fleetSelector.DeepCopy()
+
+	allocated := agonesv1.GameServerStateAllocated
+	allocatedSelector.MatchLabels[lockLabel] = "true"
+	gsa := &allocationv1.GameServerAllocation{
+		Spec: allocationv1.GameServerAllocationSpec{
+			MetaPatch: allocationv1.MetaPatch{Labels: map[string]string{lockLabel: "false"}},
+			Selectors: []allocationv1.GameServerSelector{
+				{LabelSelector: *allocatedSelector, GameServerState: &allocated},
+				{LabelSelector: fleetSelector},
+			},
+		}}
+
+	// standard allocation
+	result, err := framework.AgonesClient.AllocationV1().GameServerAllocations(fleet.ObjectMeta.Namespace).Create(ctx, gsa, metav1.CreateOptions{})
+	require.NoError(t, err)
+	require.Equal(t, string(allocationv1.GameServerAllocationAllocated), string(result.Status.State))
+
+	gs, err := framework.AgonesClient.AgonesV1().GameServers(fleet.ObjectMeta.Namespace).Get(ctx, result.Status.GameServerName, metav1.GetOptions{})
+	require.NoError(t, err)
+	require.Equal(t, allocated, gs.Status.State)
+
+	// set the label to being available again
+	_, err = framework.SendGameServerUDP(t, gs, "LABEL available true")
+	require.NoError(t, err)
+
+	// wait for the label to be applied!
+	require.Eventuallyf(t, func() bool {
+		gs, err := framework.AgonesClient.AgonesV1().GameServers(fleet.ObjectMeta.Namespace).Get(ctx, result.Status.GameServerName, metav1.GetOptions{})
+		require.NoError(t, err)
+		log.WithField("labels", gs.ObjectMeta.Labels).Info("checking labels")
+		return gs.ObjectMeta.Labels[lockLabel] == "true"
+	}, time.Minute, time.Second, "GameServer did not unlock")
+
+	// Run the same allocation again, we should get back the preferred item.
+	expected := result.Status.GameServerName
+
+	// we will run this as an Eventually, as caches are eventually consistent
+	require.Eventuallyf(t, func() bool {
+		result, err = framework.AgonesClient.AllocationV1().GameServerAllocations(fleet.ObjectMeta.Namespace).Create(ctx, gsa, metav1.CreateOptions{})
+		require.NoError(t, err)
+		require.Equal(t, string(allocationv1.GameServerAllocationAllocated), string(result.Status.State))
+
+		if expected != result.Status.GameServerName {
```

Comment:
I'm not clear on this check. This will fail if the allocation attempt succeeds but returns a different game server than the first allocation attempt, right? Does that warrant a retry or should that be considered a failure? It seems like the test could end up allocating all game servers in the fleet before coming back around and reallocating from the one from the first allocation and the test would still pass - but that isn't the intended behavior of the feature.

Reply:
Well, it has to work this way. Everything is eventually consistent (hence the change in the docs), so it should work correctly _eventually_ - the order actually doesn't matter.

We can't control when the allocation cache gets refreshed for example.

Reply:
Locally this works on first try, but I expect on the e2e server under load, it might need a few retries.